### PR TITLE
Add intersection function

### PR DESCRIPTION
**Pull request by chitang**

### CODE REVIEW

File: soql-stdlib/src/main/scala/com/socrata/soql/functions/SoQLFunctions.scala [hunk 1]
```diff
@@ -61,6 +61,8 @@ object SoQLFunctions {
     Seq(VariableType("a")), Seq.empty, FixedType(SoQLMultiPolygon), isAggregate = true)
   val Intersects = f("intersects", FunctionName("intersects"), Map("a" -> GeospatialLike, "b" -> GeospatialLike),
     Seq(VariableType("a"), VariableType("b")), Seq.empty, FixedType(SoQLBoolean))
+  val Intersection = f("intersection", FunctionName("intersection"), Map("a" -> GeospatialLike, "b" -> GeospatialLike),
+    Seq(VariableType("a"), VariableType("b")), Seq.empty, VariableType("a"))
```
[the remark this answers is not in the document]
Comment (chitang) [reply soql-stdlib/src/main/scala/com/socrata/soql/functions/SoQLFunctions.scala]: Yeah, postgis functions just take geometry.  It isn't going to cover all the combinations.  Not sure what the best compromise is.

[the remark this answers is not in the document]
Comment (bewithjonam) [reply soql-stdlib/src/main/scala/com/socrata/soql/functions/SoQLFunctions.scala]: @chitang @rjmac Hello there... 

If we create a variant of the function for each cases, then we might need to cover all the pair combinations of (Point, MultiPoint, Line, MultiLine, Polygon, MultiPolygon).

Suggestion:
```
f("line_box_intersection", FunctionName("line_box_intersection"), Map("a" -> SOQLLine, "b" -> RealNumLike, "o" -> SOQLMultiLine),
    Seq(VariableType("a"), VariableType("b"), VariableType("b"), VariableType("b"), VariableType("b")), Seq.empty, FixedType("o"))
f("multi_line_box_intersection", FunctionName("multi_line_box_intersection"), Map("a" -> SOQLMultiLine, "b" -> RealNumLike),
    Seq(VariableType("a"), VariableType("b"), VariableType("b"), VariableType("b"), VariableType("b")), Seq.empty, FixedType("a"))

f("polygon_box_intersection", FunctionName("polygon_box_intersection"), Map("a" -> GeospatialLike, "b" -> RealNumLike, "o" -> SOQLMultiPolygon),
    Seq(VariableType("a"), VariableType("b"), VariableType("b"), VariableType("b"), VariableType("b")), Seq.empty, FixedType("o"))

f("multi_polygon_box_intersection", FunctionName("multi_polygon_box_intersection"), Map("a" -> SOQLMultiPolygon, "b" -> RealNumLike),
    Seq(VariableType("a"), VariableType("b"), VariableType("b"), VariableType("b"), VariableType("b")), Seq.empty, FixedType("a"))

```
Usage: 
```
www.domain.com/resource/four-four.geojson?$query=select line_box_intersection(line_column, 0, 0, 10, 10)
=> Returns multilines
www.domain.com/resource/four-four.geojson?$query=select multi_line_box_intersection(multi_line_column, 0, 0, 10, 10)
=> Returns multilines

www.domain.com/resource/four-four.geojson?$query=select polygon_box_intersection(polygon_column, 0, 0, 10, 10)  
=> Returns multipolygons
www.domain.com/resource/four-four.geojson?$query=select multi_polygon_box_intersection(multi_polygon_box_intersection, 0, 0, 10, 10)
=> Returns multipolygons
```

Depending on the input geometries, line_box_intersection/multi_line_box_intersection
returns  either
* line (can we cast it to MultiLines in the soql-postgres-adapter)
* multi-line 
* ```GEOMETRYCOLLECTION EMPTY``` Not sure about this. Did some digging and found the geojson codec and doesn't seem we handle this. 

[socrata thirdparty JtsCodecs](https://github.com/socrata-platform/socrata-thirdparty-utils/blob/231364353b260e80dfb231de2645391c1621a4b0/core/src/main/scala/com/socrata/thirdparty/geojson/JtsCodecs.scala)

[the remark this answers is not in the document]
Comment (chitang) [reply soql-stdlib/src/main/scala/com/socrata/soql/functions/SoQLFunctions.scala]: Do we still need point, mpoint, etc?  What if we add geometry type?

[the remark this answers is not in the document]
Comment (bewithjonam) [reply soql-stdlib/src/main/scala/com/socrata/soql/functions/SoQLFunctions.scala]: Our problem now is: 
* If there is a large shape/line that runs through say n tiles, we are fetching that entire shape n times. 
* If a shape covers entire USA, and I am looking at Austin Texas, our map is fetching the entire shape.

For points, we are fetching points within a tile using within_box. Intersection not required.

For multipoints with high cardinality, we might face the above mentioned problems. This is not a frequent case, I guess.

**Pull request closed without merging**